Gruntfile.js

```diff
@@ -340,29 +368,69 @@ module.exports = (grunt) => {
+      [['backend/**/*.js', 'shared/**/*.js'], [eslint, 'backend:relaunch']],
       [['frontend/**/*.html'], ['copy']],
-      [['frontend/**/*.js'], ['exec:eslint']],
-      [['frontend/views/**/*.vue'], ['exec:eslint', 'exec:puglint', 'exec:stylelint']],
-      [['backend/**/*.js', 'shared/**/*.js'], ['exec:eslint', 'backend:relaunch']],
-      [['Gruntfile.js'], ['exec:eslint']]
+      [['frontend/**/*.js'], [eslint]],
+      [['frontend/assets/{fonts,images}/**/*'], ['copy']],
+      [['frontend/assets/style/**/*.scss'], [stylelint]],
+      [['frontend/assets/svgs/**/*.svg'], []],
+      [['frontend/views/**/*.vue'], [puglint, stylelint, eslint]]
     ].forEach(([globs, tasks]) => {
       globs.forEach(glob => {
-        browserSync.watch(glob, { ignoreInitial: true }, async (eventTypeName, filePath) => {
-          grunt.log.writeln(chalk`{green file event:} '${eventTypeName}' {green detected on} ${filePath}`)
-          // These tasks will run concurrently with the incremental rebuild.
-          grunt.task.run(tasks)
+        browserSync.watch(glob, { ignoreInitial: true }, async (eventName, filePath) => {
+          grunt.log.writeln(chalkFileEvent(eventName, filePath))
+
+          if (eventName === 'add' || eventName === 'change') {
+            // Read and lint the changed file.
+            const code = await fs.promises.readFile(filePath, 'utf8')
+            const linters = tasks.filter(task => typeof task === 'object')
+            const lintingStartMs = Date.now()
+
+            await Promise.all(linters.map(linter => linter.lintCode(code, filePath)))
+              // Don't crash the Grunt process on lint errors.
+              .catch(() => {})
+
+            // Log the linting time, formatted with Chalk.
+            grunt.log.writeln(chalkLintingTime(Date.now() - lintingStartMs, linters, [filePath]))
+          }
+
+          if (eventName === 'change' || eventName === 'unlink') {
+            const extension = path.extname(filePath)
+
+            // Remove the corresponding plugin cache entry, if any.
+            if (extension === '.js') {
+              flowRemoveTypesPluginOptions.cache.delete(filePath)
+            } else if (extension === '.svg') {
+              svgInlineVuePluginOptions.cache.delete(filePath)
+            } else if (extension === '.vue') {
+              vuePluginOptions.cache.delete(filePath)
+            }
+            // Clear the whole Vue plugin cache if a Sass or SVG file was
+            // changed since some compiled Vue files might include it.
+            if (['.sass', '.scss', '.svg'].includes(extension)) {
+              vuePluginOptions.cache.clear()
+            }
+          }
           // Only rebuild the relevant entry point.
           if (filePath.startsWith(serviceWorkerDir)) {
             await buildServiceWorkers.run()
           } else {
             await buildMain.run()
           }
+          grunt.task.run(tasks.filter(task => typeof task === 'string'))
```

Comment:
Hmm, previously this line would run before this `if` block. Should it be moved up?

Reply:
I've moved this line after the `if` block so that tasks like `backend:relaunch` do not delay rebuilds and only run if the rebuild was successful

Reply:
Although I might have made a mistake here. If a `./backend` file was changed, perhaps we want the backend server to reload before the browser refreshes